[IMP] website: Redesign and reorder the top menu

Brieuc-brd opened the pull request:
This PR adapts the design and reorganizes the elements of the systray menu to make it more readable and user-friendly.

Also, when we were in a translated page, we had two buttons in the systray menu: "**translate**" and "**or edit on master**".
To simplify the navigation, this PR now uses a Dropdown to group these two actions together.

Requires:
- https://github.com/odoo/enterprise/pull/55400

task-3626064

Co-Authored-By: Brieuc-brd <brd@odoo.com>
Co-Authored-By: Serge Bayet <seba@odoo.com>

---
| Before | After |
|--------|--------|
| ![Capture d’écran 2024-01-26 à 14 13 12](https://github.com/odoo/odoo/assets/80679690/a0d889e2-9f95-4dde-a10d-39ba82131821) | ![Capture d’écran 2024-02-15 à 13 04 56](https://github.com/odoo/odoo/assets/80679690/ccfce35e-1a69-4c57-a473-73f5ef275250) |
| ![Capture d’écran 2024-01-26 à 14 13 04](https://github.com/odoo/odoo/assets/80679690/427e3416-fb3e-40e7-91a8-59d6f3179455) | ![Capture d’écran 2024-03-08 à 11 15 30](https://github.com/odoo/odoo/assets/80679690/1d30ba9c-763c-4fc6-9d85-e521f9815d91) | 

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/151356).

SergeBayet commented:
@Brieuc-brd Hi! Could you tell me if my commit is good for you?

pierre-pde commented:
@Brieuc-brd  & @SergeBayet ,
 I had a look and this looks good to me :)

Brieuc-brd commented:
Looks good @SergeBayet, thanks ! 👍 
I'll rebase the branch :)

Brieuc-brd commented:
Hello @odoo/rd-website 👋 
These PRs (Community & Enterprise) are ready for tech. review, could you take a look at it ? :)

pierre-pde commented:
@qsm-odoo , the design in this PR is the chosen one :) 
I'm glad you like it.
Regarding [Linkedin post](https://www.linkedin.com/posts/fpodoo_odoo-activity-7155768323064946688-ybp_?utm_source=share&utm_medium=member_desktop), here is the context :D  
![image](https://github.com/odoo/odoo/assets/75257428/9c03f502-76b2-4ca4-85ba-2bbd627a14bb)

rdeodoo commented:
from @pierre-pde 

> @qsm-odoo , the design in this PR is the chosen one :)
I'm glad you like it.

rdeodoo reviewed (changes requested):
Community is never the top priority design wise but here it's really not beautiful imo
![image](https://github.com/odoo/odoo/assets/30048408/c8463476-bacb-4172-8912-450634b74399)
Could you add the same margin at least as in the enterprise?

And maybe add gray too on Edit or something, but those 2 purpose colors for New and Edit are not really nice it feels like.

other than that, nice :+1: 

Even if I am not sure I wouldn't have preferred something in between what we had and this PR, but that's subjective and this seems to please everyone :partying_face: 
Eg, I don't like the bar between the publish and the user icon, I am not sure why we need it, there is not many items or space taken by the whole navbar:
![image](https://github.com/odoo/odoo/assets/30048408/443b809e-0e43-4f6b-ac11-aacb4d5eddda)
I wonder if you did not add it because of the language switcher, which is always considered in the screenshot of this PR and FP message on linkedin, while 95% of the DBs don't have it shown as they don't have multiple website, so all comparison should have been done without it imo

Brieuc-brd commented:
Hello @rdeodoo 👋 
I've adapted colors and spaces for Community 🙂.

> I don't like the bar between the publish and the user icon, I am not sure why we need it, there is not many items or space taken by the whole navbar

The aim is to separate the user icon from the website buttons.
Without this border, the user may think that the user icon is linked to the website actions.

pierre-pde commented:
@rdeodoo , to save you some time;
![image](https://github.com/odoo/odoo/assets/75257428/783f8d27-b636-416f-a405-0a3549deb4ac)
@Brieuc-brd , thanks for the change, indeed it's better :) 
I agree that it's nice to separate website options & user profile.

rdeodoo commented on addons/website/static/src/systray_items/edit_website.xml:
```diff
@@ -1,14 +1,26 @@
+                <span t-out="label"/>
+            </button>
+            <t t-set-slot="content">
+                <DropdownItem onSelected="() => this.startTranslate()" class="'o_translate_website_dropdown_item'">
+                    Translate (<t t-out="this.websiteService.currentWebsite.metadata.lang"/>)
+                </DropdownItem>
+                <DropdownItem onSelected="() => this.startEdit()" class="'o_edit_website_dropdown_item'">
+                    Edit (<t t-out="this.websiteService.currentWebsite.metadata.defaultLang"/>)
+                </DropdownItem>
```
![image](https://github.com/odoo/odoo/assets/30048408/2e18b190-256e-4b4f-a8cc-4a12f286d0b3)

Lang codes are IMO not easy to understand (and a bit ugly), wouldn't it be better with their name?
![image](https://github.com/odoo/odoo/assets/30048408/c2312927-790e-4293-af4e-cb67e4552a45)

Maybe could be even nicer with some quick css win, probably not really needed tho
![image](https://github.com/odoo/odoo/assets/30048408/50c3782a-a3f1-4b1a-8fc7-c887f2d6126a)

Brieuc-brd replied:
Yes, indeed it would be better 👍. Although by using only the name, there will be an issue when a website will use language “variants”, such as “English US” and “English AU“. The rendering will be
- “Edit (English)”
- “Translate (English)”.

Why not use the same rendering as the language selector? Something like this : 
![Capture d’écran 2024-03-07 à 11 29 58](https://github.com/odoo/odoo/assets/80679690/f0d2f2dd-ac29-4c6e-8fc7-a161b486d6a2)

If so, @SergeBayet do you think it's possible to achieve this result ? could you have a look at it ?
I can change the style afterwards 🙂

rdeodoo replied:
That would be really better indeed

SergeBayet replied:
I'm on it :)

SergeBayet replied:
@rdeodoo @Brieuc-brd It's done. Could you have a look? thanks

Brieuc-brd replied:
Thanks @SergeBayet, looks good ! 👍

rdeodoo commented on addons/website/static/src/systray_items/mobile_preview.xml:
```diff
@@ -1,9 +1,8 @@
-         t-att-class="{ 'o_mobile_preview_active': this.state.isMobile }">
-        <a href="#" accesskey="v" class="o_nav_entry"
+    <div class="o_mobile_preview o_menu_systray_item d-none d-md-flex">
+        <a href="#" accesskey="v" class="o_nav_entry mx-1 px-3 is_full_height" t-att-class="{ 'o_mobile_preview_active': this.state.isMobile }"
```
why does `o_mobile_preview_active` need to be moved from the `div.o_mobile_preview` to its inner `a`?

Brieuc-brd replied:
My mistake, I’ve put it back 👍 .

rdeodoo reviewed (changes requested):
First one, mostly some questions and a few things to change.

For tracking purpose, here are some before/after that client will mainly see (no website switcher for 99.9% of DBs and lot of website):

| where | before | after |
| - | - | - |
| /shop  community| ![image](https://github.com/odoo/odoo/assets/30048408/3479ce08-9e8b-44bb-87e1-4ab762805ebc) | ![image](https://github.com/odoo/odoo/assets/30048408/58088d5c-dd3e-473a-804f-5dbf111bc6ab) |
| /shop enterprise |  ![image](https://github.com/odoo/odoo/assets/30048408/ce4c71b4-3352-4bd9-8805-5bc1ab1df537) | ![image](https://github.com/odoo/odoo/assets/30048408/bbc3c63c-085f-4ed4-9626-7d4fc4707266) |
| /product enterprise | ![image](https://github.com/odoo/odoo/assets/30048408/69af6135-a6be-4fbb-abd4-46983244950f) | ![image](https://github.com/odoo/odoo/assets/30048408/0d2ff99c-2589-4400-89b1-10db5c43491e) |
| /product community | ![image](https://github.com/odoo/odoo/assets/30048408/2da83107-05e3-4f61-a10b-87e42232b98a) | ![image](https://github.com/odoo/odoo/assets/30048408/ff4c085d-8c14-4be6-8651-bfe24d830cd4)
| /page community | ![image](https://github.com/odoo/odoo/assets/30048408/fe3bd10e-8130-4938-8f79-99c31c9a1efa) | ![image](https://github.com/odoo/odoo/assets/30048408/294770df-c3ae-4950-8b1b-86ea8341d863) | 
| /page enterprise | ![image](https://github.com/odoo/odoo/assets/30048408/07125395-2c4c-42fb-9f25-b3f3b6fda595) | ![image](https://github.com/odoo/odoo/assets/30048408/bf232a8e-32bf-4d0f-8d06-4cf2f6b14e6f)

Nice job overall :+1: 
I really like:
- the icon removal to declutter a bit the UI (except for the Edit in backend)
- the smaller toggle icon
- the extra color(s) on Edit and new

I am still really not a fan of the separator, but I am the only one so just ignore me about that part ;-)

pierre-pde commented on addons/website/static/src/systray_items/edit_in_backend.xml:
```diff
@@ -2,9 +2,10 @@
-        <a href="#" class="o_nav_entry" accesskey="e" t-on-click="editInBackend">
-            <span class="fa fa-cog"/>
-            <span class= "d-none d-md-block ms-2" t-esc="state.mainObjectName"/>
+        <span class="o_website_systray_separator d-none d-md-block w-0 my-auto border-start"/>
+        <a href="#" class="o_nav_entry o_nav_edit_in_backend btn d-flex align-items-center mx-1 border-0 rounded-0 px-2 px-md-3 is_full_height" accesskey="e" t-on-click="editInBackend">
+            <span class="fa fa-cog d-block d-md-none"/>
+            <span class= "d-none d-md-block" t-esc="state.mainObjectName"/>
```
@rdeodoo & @qsm-odoo , we already discussed this a lot with Fabien and others.
I personally believe that "Product" will work like this.
I would say let's go and worst case scenario you change it later :) 
cc @Brieuc-brd

Brieuc-brd commented on addons/web/static/src/webclient/navbar/navbar.scss:
```diff
@@ -26,7 +26,7 @@
     // --------------------------------------------------------------------------
 
-    .o_nav_entry, .dropdown-toggle {
+    .o_nav_entry, .dropdown-toggle:not(.o-website-dropdown-toggle-custo) {
```
Done 👍 , I replaced it with `.o-dropdown-toggle-custo`

Brieuc-brd commented on addons/website/static/src/components/navbar/navbar.scss:
```diff
@@ -1,4 +1,23 @@
+        margin-bottom: -$o-navbar-padding-v;
+        padding-top: $o-navbar-padding-v;
+        padding-bottom: $o-navbar-padding-v;
+    }
```
Using this global selector will also have an impact on unwanted buttons, such as the "new elements" button (new page, new blog post, new event, …).
So I had to specify the selector like this, what do you think ?
https://github.com/odoo/odoo/pull/151356/files#diff-fba2db69aa3def89b15e5c275d5eead31420b4996c8ca04be59be5ee952abfc0R13

Brieuc-brd commented on addons/website/static/src/services/website_service.js:
```diff
@@ -227,7 +229,8 @@ export const websiteService = {
             },
             async fetchWebsites() {
-                websites = [...(await orm.searchRead('website', [], ['domain', 'id', 'name']))];
+                websites = [...(await orm.searchRead('website', [], ['domain', 'id', 'name', 'default_lang_id']))];
+                defaultLang = (await orm.searchRead("res.lang", [["id", "=", websites[0]["default_lang_id"][0]]], ["code"]))[0]["code"];
```
Hey @SergeBayet  could you have a look at this ☝️ 🙂  ?

Brieuc-brd commented on addons/website/static/src/services/website_service.js:
```diff
@@ -227,7 +229,8 @@ export const websiteService = {
             },
             async fetchWebsites() {
-                websites = [...(await orm.searchRead('website', [], ['domain', 'id', 'name']))];
+                websites = [...(await orm.searchRead('website', [], ['domain', 'id', 'name', 'default_lang_id']))];
+                defaultLang = (await orm.searchRead("res.lang", [["id", "=", websites[0]["default_lang_id"][0]]], ["code"]))[0]["code"];
```
@SergeBayet And also this ☝️ 😄 ?

Brieuc-brd commented on addons/website/static/src/systray_items/edit_website.js:
```diff
@@ -61,7 +78,7 @@ class EditWebsiteSystray extends Component {
 export const systrayItem = {
     Component: EditWebsiteSystray,
-    isDisplayed: (env) => env.services.website.currentWebsite.metadata.editable,
+    isDisplayed: (env) => env.services.website.currentWebsite.metadata.editable || (env.services.website.currentWebsite && env.services.website.currentWebsite.metadata.translatable),
```
Since I combined `translate_website.js` and `edit_website.js` in a same file, I thought we had to keep it as it was :
https://github.com/odoo/odoo/pull/151356/files#diff-b038a51402b0f40d654996cea971c4aafe939f9571f858ec14ca957969685243L32
I've updated the code and it seems to work fine 👍

rdeodoo commented on addons/website/static/src/js/tours/tour_utils.js:
```diff
@@ -195,6 +195,29 @@ function clickOnEditAndWaitEditMode(position = "bottom") {
+ */
+function clickOnEditAndWaitEditModeInTranslatedPage(position = "bottom") {
+    return [{
+        content: markup(_t("<b>Click Edit</b> dropdown")),
+        trigger: ".o_edit_website_container button",
+        position: position,
+    }, {
+        content: markup(_t("<b>Click Edit</b> to start designing your homepage.")),
```
@odoo/rd-security While this PR is getting reviewed, would you mind already overriding the `ci/security` triggered by those 2 `markup()`?
Those are safe, we just inject our own hardcoded `<b/>`.

Brieuc-brd commented:
Hello @rdeodoo 👋 

@SergeBayet  and I have updated and rebased the branch.
The diffs : 
- https://github.com/odoo/odoo/compare/2f4ae4c616faf8c36fe79037ad7185733db7ad5b..94977c28d98fe6f8118d24557d103dec874a2754
- https://github.com/odoo/odoo/compare/94977c28d98fe6f8118d24557d103dec874a2754..920c09635b4767186a17db4935e31d02feced3a8

Could you take a look at it? :)

rdeodoo commented on addons/website/static/src/js/content/inject_dom.js:
```diff
@@ -46,6 +46,12 @@ document.addEventListener('DOMContentLoaded', () => {
+    const website_default_lang_name = session.website_default_lang_name;
+    if (website_default_lang_name) {
+        htmlEl.dataset.defaultLangName = website_default_lang_name;
+        htmlEl.dataset.langName = session.website_lang_name;
```
I suggested `get_frontend_info` because I thought we could just do `session.xx` where we needed it (in the website_service), but since it's not possible (because it's the __session_info from the backend document) and you then had to "forward" / "pass" it from the iframe to the backend by setting it to the iframe html data attributes, it was better to just add it directly in those data attributes, which is done in the `website.layout` with `t-set="html_data"`.

rdeodoo reviewed (changes requested):
I applied the changes requested myself here:

But actually, there is still a small UI glitch, could you have a look?
![image](https://github.com/odoo/odoo/assets/30048408/b8df9c63-3021-4626-a916-bac07fa9f1e3)
The `o_mobile_menu_toggle` is broken in website.

Then I'll merge it :+1:

SergeBayet commented on addons/website/static/src/js/content/inject_dom.js:
```diff
@@ -46,6 +46,12 @@ document.addEventListener('DOMContentLoaded', () => {
         htmlEl.dataset.country = country;
     }
 
+    const website_default_lang_name = session.website_default_lang_name;
```
@rdeodoo Sorry for that I was distracted by the session properties which are snake-cased. That's in order now.

robodoo commented:
Merge method set to rebase and fast-forward.
